module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,688 @@
+	header, err := jobs.JobToBlock(job)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to convert job to block: %w", err))
+	}
+
+	request := &jobs.BlockEntry{
+		BlockID: header.ID(),
+		Height:  header.Height,
+	}
+
+	err = e.processSealedHeight(ctx, request)
+	if err == nil {
+		complete()
+	}
+
+	// errors are thrown as irrecoverable errors except context cancellation, and invalid blobs
+	// invalid blobs are logged, and never completed, which will halt downloads after maxSearchAhead
+	// is reached.
+}
+
+// processSealedHeight downloads ExecutionData for the given block height.
+// If the download fails, it will retry forever, using exponential backoff.
+func (e *executionDataRequester) processSealedHeight(ctx irrecoverable.SignalerContext, request *jobs.BlockEntry) error {
+	backoff := retry.NewExponential(e.config.RetryDelay)
+	backoff = retry.WithCappedDuration(e.config.MaxRetryDelay, backoff)
+	backoff = retry.WithJitterPercent(15, backoff)
+
+	attempt := 0
+	return retry.Do(ctx, backoff, func(context.Context) error {
+		if attempt > 0 {
+			e.metrics.FetchRetried()
+		}
+		attempt++
+
+		// download execution data for the block
+		err := e.processFetchRequest(ctx, request)
+
+		// don't retry if the blob was invalid
+		if isInvalidBlobError(err) {
+			return err
+		}
+
+		return retry.RetryableError(err)
+	})
+}
+
+func (e *executionDataRequester) processFetchRequest(ctx irrecoverable.SignalerContext, request *jobs.BlockEntry) error {
+	logger := e.log.With().
+		Str("block_id", request.BlockID.String()).
+		Uint64("height", request.Height).
+		Logger()
+	logger.Debug().Msg("processing fetch request")
+
+	result, err := e.results.ByBlockID(request.BlockID)
+
+	// By the time the block is sealed, the ExecutionResult must be in the db
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to lookup execution result for block: %w", err))
+	}
+
+	executionData, err := e.fetchExecutionData(ctx, result.ExecutionDataID, request.Height)
+
+	if isInvalidBlobError(err) {
+		// This means an execution result was sealed with an invalid execution data id (invalid data).
+		// Eventually, verification nodes will verify that the execution data is valid, and not sign the receipt
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("HALTING REQUESTER: invalid execution data found")
+
+		e.metrics.Halted()
+		return err
+	}
+
+	// Some or all of the blob was missing or corrupt. retry
+	if isBlobNotFoundError(err) || errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
+		logger.Error().Err(err).Msg("failed to get execution data for block")
+
+		return err
+	}
+
+	// Any other error is unexpected
+	if err != nil {
+		logger.Error().Err(err).
+			Str("execution_data_id", result.ExecutionDataID.String()).
+			Msg("unexpected error fetching execution data")
+
+		ctx.Throw(err)
+	}
+
+	logger.Debug().Msg("Fetched execution data")
+
+	e.executionDataCache.Add(request.Height, executionData)
```

Comment:
I don't think the usage of the DataCache is safe.

By exposing the `Add` and `Remove` method, the executionDataCache relies on the caller to manage the cache items. The DataCache itself is just like a simple dictionary. That's very little responsibility for even have the cache module. 

And more importantly, if we are adding a cache layer for storage, it's important to ensure the cache see all the updates to the storage. 

Usually it could be done by wrapping the storage API with the cache API, just like [the cache module](https://github.com/onflow/flow-go/blob/petera/execution-data-requester/storage/badger/headers.go#L78-L82).

So I would suggest to create the `executionDataCache` by wrapping the execution data service. And let `execution data requester` also use the `executionDataCache` instead.

Reply:
I've removed the caching from this PR